src/ipc/server.rs

```diff
@@ -108,42 +107,36 @@ fn on_new_ipc_client(state: &mut State, stream: UnixStream) {
 
 async fn handle_client(ctx: ClientCtx, stream: Async<'_, UnixStream>) -> anyhow::Result<()> {
     let (read, mut write) = stream.split();
+    let mut buf = String::new();
 
-    // note that we can't use the stream json deserializer here
-    // because the stream is asynchronous and the deserializer doesn't support that
-    // https://github.com/serde-rs/json/issues/575
+    // Read a single line to allow extensibility in the future to keep reading.
+    BufReader::new(read)
+        .read_line(&mut buf)
+        .await
+        .context("error reading request")?;
 
-    let mut lines = BufReader::new(read).lines();
+    let request = serde_json::from_str(&buf)
+        .context("error parsing request")
+        .map_err(|err| err.to_string());
+    let requested_error = matches!(request, Ok(Request::ReturnError));
 
-    let line = match lines.next().await.unwrap_or(Err(io::Error::new(io::ErrorKind::UnexpectedEof, "Unreachable; BufReader returned None but when the stream ends, the connection should be reset"))) {
-        Ok(line) => line,
-        Err(err) if err.kind() == io::ErrorKind::ConnectionReset => return Ok(()),
-        Err(err) => return Err(err).context("error reading line"),
-    };
-
-    let reply: Reply = serde_json::from_str(&line)
-        .map_err(|err| format!("error parsing request: {err}"))
-        .and_then(|req| process(&ctx, req));
+    let reply = request.and_then(|request| process(&ctx, request));
 
     if let Err(err) = &reply {
-        warn!("error processing IPC request: {err:?}");
+        if !requested_error {
```

Comment:
Do we want to handle this error flow differently? It's not really problematic to print it; nobody's gonna spam `niri msg request-error` and then complain that the log output contains the same message; conversely, it's reasonable that someone may want to see what it looks like on the compositor side handling an error in which case this check somewhat obscures it.

Reply:
Well, no hard opinions here, but in my head this request is aimed much more at the client side, so as far as the server is concerned it shouldn't print a warning.